### DISCUSSION TITLE
Question on using ROS message as a tree precondition

### DISCUSSION THREAD

**Discussion by quantumxt** [BehaviorTree/BehaviorTree.CPP#639] [Q&A]
Hi there,

We are currently working on a project that uses BehaviorTree.CPP 4.2, and we are considering using the scripting functionality to decide whether to run a particular tree. For example, we have a custom ROS message:
```
string MODE_STOP = "STOP"
string MODE_RUN = "RUN"
string mode
```
And we have a tree that looks something like this:
```xml
...
 <BehaviorTree ID="MainTree">
            <Sequence>
                <ModeCheck name="Mode Checker"
                    out_bt_mode="{run_mode}"
                    />
                <SubTree ID="RunTree"
                   _skipIf="!(run_mode.mode=='RUN')/>
...
```
Where it would perform a check on this custom msg outputted by `ModeCheck` plugin, and would skip the tree if `run_mode.mode` is not `MODE_RUN`.

Is this method of using scripting for custom ROS msg possible? Any insights or feedback would be appreciated!

**Comment by facontidavide** [comment]
First, upgrade to 4.3... a lot of bugs were fixed recently!

Actually, that code is not supposed to work: `run_mode.mode`. 
You should use `_skipIf="run_mode !='RUN'"`

I suppose that **ModeCheck** is the node that reads your ROS topic and based on the value received, updates `run_mode` accordingly.

Note that this behavior is not reactive, i.e. RunTree will not be stopped if a "STOP" messages is published

**Reply by quantumxt** [reply]
Thank you for the reply @facontidavide!

For further context, the `out_bt_mode` would output a custom ROS2 msg structure as mentioned previously. Let's say this custom message, `RunMode`, is stored as `robo/msg/RunMode.msg`. We are exploring whether we can perform a comparison with the built-in fields of the ROS2 message instead of hardcoding them. (Preferably `run_mode.mode==robo::msg::RunMode::MODE_RUN` as a preconditional check instead of doing the check inside the plugin.)

Does this mean we are not able to check the internal fields as a condition (example, checking `run_mode.mode`)? Or we would have to extract the field internally & publish it out as a blackboard variable for it to be used as a precondition check?

**Reply by facontidavide** [reply]
> Does this mean we are not able to check the internal fields as a condition

Correct, this is **not** supported by the scripting language and there is no plan to support it in the future.

**Reply by asasine** [reply]
FWIW this would be an amazing feature. A ton of our custom action nodes are for simple things like expanding the fields of a struct/message into output ports. I understand it's difficult and not super feasible since the field accessors are compiled C++ but the scripting isn't, but just putting my support behind the idea.

A contrived example, where I want to write a condition on the yaw in a `geometry_msgs::msg::PoseStamped`:

```xml
<Sequence>
  <ReadPoseStamped pose_stamped="{pose_stamped}" pose="{pose}" header="{header}" />
  <ReadPose pose="{pose}" position="{position}" orientation="{orientation}" />
  <ReadPosition position="{position}" x="{x}" y="{y}" z="{z}" />
  <QuaternionToRPY quaternion="{orientation}" roll="{roll}" pitch="{pitch}" yaw="{yaw}" />

  <!-- now I can finally do something with yaw --->
  <YawCondition yaw="{yaw}" />
</Sequence>
```

`ReadPoseStamped`, `ReadPose`, and `ReadPosition` would be replaceable by scripting if this feature existed.

```xml
<Sequence>
  <QuaternionToRPY quaternion="{pose_stamped.pose.orientation}" yaw="{yaw}" />
  <YawCondition yaw="{yaw}" />
</Sequence>
```

**Comment by facontidavide** [comment]
Sorry, but not going to happen for the time being
